### PR TITLE
feat(lsp): simple diagnostics

### PR DESCRIPTION
**Pull request by Natoandro**
### Describe your change

Adds a simple LSP implementation for diagnostics.

### Motivation and context

Better DX, catch potential errors while editing the typegraph module.

### Checklist

- [x] The change come with new or modified tests
- [x] Hard-to-understand functions have explanatory comments
- [ ] End-user documentation is updated to reflect the change

### CODE REVIEW

File: dev-tools/ts-language-server/src/server.ts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+documents.onDidChangeContent((change) => {
+  validateTextDocument(change.document);
+});
+
+async function validateTextDocument(textDocument: TextDocument): Promise<void> {
+  // In this simple example we get the settings for every validate run.
+  const settings = await getDocumentSettings(textDocument.uri) ??
+    defaultSettings;
+
+  // The validator creates diagnostics for all uppercase words length 2 and more
```
Comment (Yohe-Am) [inline dev-tools/ts-language-server/src/server.ts:145]: issue: appears to be be a stale comment.

File: dev-tools/ts-language-server/src/server.ts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  const diagnostics: Diagnostic[] = [];
+  while ((m = pattern.exec(text)) && problems < settings.maxNumberOfProblems) {
+    problems++;
+    const diagnostic: Diagnostic = {
+      severity: DiagnosticSeverity.Error,
+      range: {
+        start: textDocument.positionAt(m.index),
+        end: textDocument.positionAt(m.index + m[0].length),
+      },
+      message: `${m[0]} is all uppercase.`,
```
Comment (Yohe-Am) [inline dev-tools/ts-language-server/src/server.ts:160]: Oh wait, is this function just a stub impl as of now?

File: dev-tools/vscode-metatype-support/src/extension.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    path.join("..", "ts-language-server"),
+  );
+  const serverOptions: ServerOptions = {
+    run: {
+      command: "typegraph-ts-server",
+      args: ["--stdio"],
+      transport: TransportKind.stdio,
+    },
+    debug: {
+      command: "deno",
```
Comment (Yohe-Am) [inline dev-tools/vscode-metatype-support/src/extension.ts:24]: question: Any reason you decided to write the lsp in deno? now that `meta-cli` embeds its own deno, the deno binary itself will be a unique requirement of the lang server. If the vs-code environment provides node by default, might be better in that sense.

File: dev-tools/ts-language-server/src/server.ts [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+async function validateTextDocument(textDocument: TextDocument): Promise<void> {
+  // In this simple example we get the settings for every validate run.
+  const settings = await getDocumentSettings(textDocument.uri) ??
+    defaultSettings;
+
+  // The validator creates diagnostics for all uppercase words length 2 and more
+  const text = textDocument.getText();
+  const pattern = /\.apply\(/g;
+  let m: RegExpExecArray | null;
```
Comment (afmika) [inline dev-tools/ts-language-server/src/server.ts:148]: Correct me if I am wrong but wouldn't this retrieve all `.apply` expressions even custom ones (eg. user defined utility classes methods). Also What if .apply is inside a string or a comment ?

File: dev-tools/ts-language-server/tests/expose_analysis_test.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+      runtime: res.runtime.text,
+      generator: res.generator,
+    };
+  });
+
+  assertObjectMatch(analysisResult[0], {
+    name: "add",
+    runtime: "python",
+    generator: "fromLambda",
+  });
```
Comment (afmika) [inline dev-tools/ts-language-server/tests/expose_analysis_test.ts:45]: Nice!

File: dev-tools/ts-language-server/src/server.ts (hunk 4 above)
Comment (Natoandro) [reply dev-tools/ts-language-server/src/server.ts:148]: It was just the initial tests. Semantic analysis is still a WIP.

File: dev-tools/ts-language-server/src/analysis/typescript-semantic/scope.ts [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+            subBlocks.push(block);
+          } else {
+            console.warn("skipped nested subblock", block);
+          }
+        }
+      }
+
+      switch (child.type) {
+        case "variable_declaration": {
+          throw new Error("todo");
```
Comment (Yohe-Am) [inline dev-tools/ts-language-server/src/analysis/typescript-semantic/scope.ts:51]: question: does this and similar other statements crash the server? If so, might be best to convert to a log statement or something.

**Review by Yohe-Am — APPROVED** [review]

File: dev-tools/ts-language-server/src/analysis/typescript-semantic/scope.ts (hunk 6 above)
Comment (Natoandro) [reply dev-tools/ts-language-server/src/analysis/typescript-semantic/scope.ts:51]: Yes. You're right. This would crash the server.

**Comment by Yohe-Am** [comment]
@Natoandro I think what's happening [here](https://github.com/metatypedev/metatype/actions/runs/7456787642/job/20287978695) is that the `dev/lock.ts` script is touching some `Cargo.toml` files from the tree sitter packages pulled in by `pnpm`. Looking at the rules in `dev/lock.yml` for `Cargo.toml` files, those shouldn't really qualify so I'm not sure what's going on.

**Review by Yohe-Am — APPROVED** [review]

**Review by Yohe-Am — APPROVED** [review]

**Comment by Yohe-Am** [comment]
Ah, so it doesn't let the original author approve a PR even though a separate author pushed the most recent changes.

**Review by afmika — APPROVED** [review]

**Pull request merged**